tokio/src/net/windows/named_pipe.rs

```diff
@@ -0,0 +1,644 @@
+    ///
+    /// This corresponds to specifying [nInBufferSize].
+    ///
+    /// [nInBufferSize]: https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-createnamedpipea
+    pub fn in_buffer_size(mut self, buffer: u32) -> Self {
+        self.in_buffer_size = buffer as DWORD;
+        self
+    }
+
+    /// Create the named pipe identified by the name provided in [new] for use
+    /// by a server.
+    ///
+    /// This function will call the [CreateNamedPipe] function and return the
+    /// result.
+    ///
+    /// [new]: NamedPipeBuilder::new
+    /// [CreateNamedPipe]: https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-createnamedpipea
+    ///
+    /// ```no_run
+    /// use tokio::net::windows::NamedPipeBuilder;
+    ///
+    /// # fn main() -> std::io::Result<()> {
+    /// let builder = NamedPipeBuilder::new("\\\\.\\pipe\\mynamedpipe");
+    /// let pipe = builder.create()?;
+    /// # Ok(()) }
+    /// ```
+    pub fn create(&self) -> io::Result<NamedPipe> {
+        // Safety: We're calling create_with_security_attributes w/ a null
+        // pointer which disables it.
+        unsafe { self.create_with_security_attributes(ptr::null_mut()) }
+    }
+
+    /// Create the named pipe identified by the name provided in [new] for use
+    /// by a server.
+    ///
+    /// This is the same as [create][NamedPipeBuilder::create] except that it
+    /// supports providing security attributes.
+    ///
+    /// [new]: NamedPipeBuilder::new
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure that `attrs` points to an initialized instance of
+    /// a [SECURITY_ATTRIBUTES] structure.
+    ///
+    /// [SECURITY_ATTRIBUTES]: [winapi::um::minwinbase::SECURITY_ATTRIBUTES]
+    pub unsafe fn create_with_security_attributes(&self, attrs: *mut ()) -> io::Result<NamedPipe> {
+        let h = namedpipeapi::CreateNamedPipeW(
+            self.name.as_ptr(),
+            self.open_mode,
+            self.pipe_mode,
+            self.max_instances,
+            self.out_buffer_size,
+            self.in_buffer_size,
+            self.default_timeout,
+            attrs as *mut minwinbase::SECURITY_ATTRIBUTES,
+        );
+
+        if h == handleapi::INVALID_HANDLE_VALUE {
+            return Err(io::Error::last_os_error());
+        }
+
+        let io = mio::windows::NamedPipe::from_raw_handle(h);
+        let io = PollEvented::new(io)?;
+
+        Ok(NamedPipe { io })
+    }
+}
+
+impl fmt::Debug for NamedPipeBuilder {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let name = String::from_utf16_lossy(&self.name[..self.name.len() - 1]);
+        f.debug_struct("NamedPipeBuilder")
+            .field("name", &name)
+            .finish()
+    }
+}
+
+/// A builder suitable for building and interacting with named pipes from the
+/// client side.
+///
+/// See [NamedPipeClientBuilder::create].
+#[derive(Clone)]
+pub struct NamedPipeClientBuilder {
+    name: Arc<[u16]>,
+}
```

Comment:
Afaik you also need to set the inbound/outbound flag, and it must match the value set on the server.

Reply:
This was added all though with super obvious names (`read` and `write`) since that matches the underlying operation. This is not final, I'm just punting on naming (and adding wrapping types in general) til we have a good idea of what all of the APIs look like.

I also added a slew of test cases for permissions. One which revealed something I didn't expect:

https://github.com/tokio-rs/tokio/pull/3760/files#diff-697d0148d22589ebb6a5805054e91924230460d8038f36e74d388db00e2a35f0R78